### DISCUSSION TITLE
My ideas which can be added to roadmap ?

### DISCUSSION THREAD

**Discussion by andr1972** [pfusik/cito#54] [Ideas]
There are a lot of things, but I would like to take care of it in my free time too.
* Antlr .g4 grammar of Ć language
* cito in Ć will be uncomfortable for compiling, unless:
* first creating toci :  C# (subset) ->Ć transpiler, can encourage many (which have own C# code) to use Ć.
* plugin for Intellij Idea or Visual Studio which syntax highlight Ć and compile fore example Ć->Java->Class and run
* many other languages : this is a lot of potential work: Kotlin, Pony, Delphi Pascal, ICP ()Internet Computer smartcontracts), Ada... Rust is popular but will problems with ownership managements.

**Comment by pfusik** [comment]
> Antlr .g4 grammar of Ć language

What's the purpose?

> cito in Ć will be uncomfortable for compiling, unless:

The plan is to transpile cito to C++ and release native binaries for Windows, macOS and Linux. Basically, I'd like to avoid the dependency on .NET.

> first creating toci : C# (subset) ->Ć transpiler, can encourage many (which have own C# code) to use Ć.

Do you have this subset defined? How would you deal with object/array/string ownership?

> plugin for Intellij Idea or Visual Studio which syntax highlight Ć and compile fore example Ć->Java->Class and run

Yes, IDE/editor support is definitely needed. Also, some kind of project files - we had an initial discussion in #6. But first, I'd like to remove .NET dependency.

> many other languages : this is a lot of potential work: Kotlin, Pony, Delphi Pascal, ICP ()Internet Computer smartcontracts), Ada... Rust is popular but will problems with ownership managements.

- Delphi/Pascal - #52 
- Kotlin AFAIK interoperates smoothly with Java. Also, Android Studio transpiles Java to Kotlin. What would be the advantages of adding a Kotlin backend to cito?
- Pony, ICP, Ada - I'm not familiar with these
- Rust - it's certainly popular, but I don't believe one can automatically translate Ć to idiomatic Rust

**Comment by hovi** [comment]
Kotlin indeed interoperates very well with java, but kotlin is so much better in many ways.
Some things I personally see as big benefits:
- Some things you just cannot do in java or they are much harder to do.
- Nullability (especially since cito now natively supports it). For someone it's small thing, for me it's HUGE!
- Kotlin is not just on JVM:
I feel like kotlin has similar philosophy to your "tool for multiplatform libraries" with kotlin multiplatform. "common" code can be compiled to multiple platforms including iOS, android, java, javascript and some natives like linux. This is obviously useless to you, but it is just opening multiple ways to go. For example `cito -> kotlin -> javascript`. In ideal world I would just use cito for everything "multiplatform", however I would see practical myself to co this kotlin intermediate route simply because kotlin and it's ecosystem tools are much more advanced than cito's (well obviously, they have huge funding and lots of people supporting it, including google).
- Transpiling from Java to Kotlin sucks. For example mentioned nullabilities are often done pretty badly and very often doesn't recognize non-nullable. The code works, but is horrible to use/rewrite.
- Yes I am big kotlin lover so I am definitely biased. I have been using kotlin the most among languages in past 3 years.

**Reply by pfusik** [reply]
> Some things you just cannot do in java or they are much harder to do.

Could you give some examples?

**Reply by hovi** [reply]
Just to be clear. What I meant by  "you just cannot do" is that you cannot do it in efficient or short manner. I you can do pretty much anything in Java as well, just the amount of effort is completely different on language level, not talking about libraries.
So I mainly meant nice language constructs that you can emulate in Java by using static methods, polymorphism or just bunch of extra boilerplate code.

More relevant to languages/cito:
- [Data classes](https://kotlinlang.org/docs/data-classes.html) - nice encapsulation of "struct"-like type with premade toString, copy, destructuring, equality checks and constructors. Can be done in Java, but is incredibly verbose.
- Nullability
- [Operator overloading](https://kotlinlang.org/docs/operator-overloading.html) (Can use regular functions in java)
- Lambdas, pure functions

Probably not as relevant:
- [Extension functions](https://kotlinlang.org/docs/extensions.html) (Can use static functions in java)
- Smart casting (Has to be done manually in java)
- Inline functions
- Reified generic parameters for inline functions (there is no alternative in java, you have to pass Class as another parameter yourself)
- Coroutines
